Improved auto start and respect aspect ratio features

angeldeejay opened the pull request:
Added new features:

- Aspect ratio correction (solves [ISSUE 35](https://github.com/BlankSourceCode/obs-zoom-to-mouse/issues/35))
- Auto start zoom

angeldeejay committed:
Improved auto start and force 16:9 aspect ratio features

BlankSourceCode commented:
@angeldeejay Thanks for the PR! 

Would it be possible to (either automatically by grabbing info from OBS or manually via textboxes) allow different output aspect ratios rather than just a single toggle for 16:9? I'm pretty sure it would then also fix #31 and #30.

The auto-zoom toggle may be able to fix #36 if it saves/restores the state based on the scene but that seems like a bit of a stretch goal.

angeldeejay commented:
> @angeldeejay Thanks for the PR!
> 
> Would it be possible to (either automatically by grabbing info from OBS or manually via textboxes) allow different output aspect ratios rather than just a single toggle for 16:9? I'm pretty sure it would then also fix #31 and #30.
> 
> The auto-zoom toggle may be able to fix #36 if it saves/restores the state based on the scene but that seems like a bit of a stretch goal.

The "auto-zoom" feature I've implemented, just starts the script with OBS (without using key shortcut or anything). I've found it useful in cases when you need the zoom always active. I t not pretends to fix #36.

About the aspect ratios feature: this is an initial workaround to maintain proportions on ultrawide screens. The future goal should use the original aspect ratio from the source/device the script is trying to zoom in and adjust according to it, but I'm not able to implement something with that complexity right now due to I'm busy. (Sorry 😔).

angeldeejay committed:
Improved obs startup. Bugfix

nschurmann commented:
I do like the idea of having an auto_start feature, however I tried this PR and is not working. It seems that you need to change an option, any option, to actually zoom. Then it get's locked in. @angeldeejay

cyear commented:
> 增加新功能：
> 
> * 长宽比校正（解决[问题35](https://github.com/BlankSourceCode/obs-zoom-to-mouse/issues/35)）
> * 原文地址：https://www.youtube.com/watch?v= 1

Wow, thanks a lot, it made my script work fine!

angeldeejay commented:
I've changed it, but I'm waiting to get some time to test it properly and update the branch

cyear commented:
> I've changed it, but I'm waiting to get some time to test it properly and update the branch

May I ask if the issue of not restoring after scaling is resolved? (I have encountered this problem)?

angeldeejay commented:
A short follow up:

- Regarding this:
  > > I've changed it, but I'm waiting to get some time to test it properly and update the branch
  > 
  > May I ask if the issue of not restoring after scaling is resolved? (I have encountered this problem)?

  Please provide steps to reproduce the problem or open another issue. I'm using 2 monitors with 21:9 and 32:9 (aspect ratios) and have no issues with any of them.

- Regarding this:

  > I do like the idea of having an auto_start feature, however I tried this PR and is not working. It seems that you need to change an option, any option, to actually zoom. Then it get's locked in. @angeldeejay

  I apologize for the confusion. What I meant with "auto start feature" was I want to get "locked in" auto-zoom always, not deactivate it. Even if I'm restarting OBS, it should keep auto zooming correctly.

  Maybe It is not the feature you are looking for, and we should rename the option label.

  My proposal to get working this feature in a way that can be deactivated is to add logic to keep synced auto-lock and activation status. In this way, you can deactivate auto zooming whenever you want, and if you shutdown OBS with auto zoom activated, restore auto zoom at the moment you start OBS again. Sadly, I'm not able to do this fix right now due to I'm currently involved in too many projects, sorry

**Changes done for now:**

- Renamed "auto start" feature. Now is "lock-in". It has now a description explaining this as a tooltip.
- Leave only featured changes against the main branch (keep code easy to review)

![image](https://github.com/user-attachments/assets/6386b822-5a5f-461d-bb1b-1e0da84cd08d)

I will be in touch, for developing the features you are requesting soon

angeldeejay commented:
@BlankSourceCode please open a feature request for the aspect ratio thing and tag me. I will be glad to review it with you and understand your request. 

I can suggest providing a combo with common aspect ratios. If you let the user do whatever you want with the ratios, the script probably crashes OBS easily (I've really tried to do with textboxes)

angeldeejay committed:
Fixed locked-in, fixed aspect ratio

angeldeejay commented:
**Last update**

- Fixed bugfix on **Auto Start** start feature. It auto-zoom when OBS starts, but you can toggle zoom as always with keybinding
- Aspect ratio param replaced in favor of **Keep Shape** param. When **Keep Shape** is enabled, the zoom window keeps aspect ratio according the aspect ratio has the source in canvas. This way, you can resize the source as you want, and the zoom window will "adapts" to the size of the source dynamically. Instead of adding parameters to setting up the aspect ratio, you should enable this option and resize the source to the aspect ratio you want.  Easy peasy.

![image](https://github.com/user-attachments/assets/1c694ecb-f491-4cbd-b4cd-f31a873157aa)

![image](https://github.com/user-attachments/assets/a2897acc-0c72-42fe-b021-5e6898758934)

![image](https://github.com/user-attachments/assets/c8c19167-9609-4562-a0b4-6221a23d0ec7)

Achuan-2 commented:
It would be even better if it could automatically decide whether to Zoom according to mouse event just like [focusSee](https://focusee.canny.io/)

angeldeejay commented:
> It would be even better if it could automatically decide whether to Zoom according to mouse event just like [focusSee](https://focusee.canny.io/)

It actually does it that way. The screenshots above are just showing how the aspect ratio doesn't break the plugin with this update. Let's remember that contributors of this plugin are not focused on replicate FocuSee.
